File: first_package/script/subscriber_demo.py
#!/usr/bin/env python3

"""
This script initializes and runs a ROS 2 subscriber node using the rclpy library.

The subscriber node is defined in the `SubscriberNode` class within the `first_package` package.
It subscribes to messages on a specific topic and processes them as defined in the class implementation.

Attributes:
    rclpy (module): The ROS 2 Python client library used to interact with ROS 2 systems.

Functions:
    main(args=None): Initializes the ROS 2 system, creates and spins a subscriber node, and shuts down the system.

Usage:
    To run this script, use the following command in a terminal:
    ```
    ros2 run first_package subscriber_demo.py
    ```
"""

# ROS 2 Python client library for writing ROS 2 nodes
import rclpy

# Import the SubscriberNode class from first_package
from first_package.subscriber_interface import (
    SubscriberNode,
)


def main(args=None):
    """
    Main function to initialize the ROS 2 node and keep it alive until it's manually terminated or receives a shutdown signal.

    Args:
        args (list, optional): Command-line arguments that can be passed to the ROS 2 node. Defaults to None.

    This function performs the following steps:
    1. Initializes the ROS 2 Python client library (rclpy) with any provided arguments.
    2. Creates an instance of `SubscriberNode`, which subscribes to a specific topic.
    3. Spins (i.e., continuously processes messages on all subscribers) the node to keep it alive.
    4. Destroys the node and shuts down the ROS 2 system once the node stops spinning.
    """
    # Initialize the ROS 2 Python client library
    rclpy.init(args=args)
    # Create a new instance of SubscriberNode named 'subscriber_py'
    node = SubscriberNode("subscriber_py")
    # Keep the node alive and processing data until shutdown
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        # Log a message when the node is manually terminated
        node.get_logger().warn("Keyboard interrupt detected")
    finally:
        # Cleanly destroy the node instance
        node.destroy_node()
        # Shut down the ROS 2 Python client library
        rclpy.shutdown()


if __name__ == "__main__":
    main()  # Execute the main function when the script is run

File: first_package/src/subscriber_demo.cpp
/**
 * @file subscriber_node_main.cpp
 * @brief Example of a ROS 2 Subscriber Node in C++.
 *
 * This file demonstrates the creation and usage of a ROS 2 subscriber node
 * using the rclcpp library. The SubscriberNode class is utilized here to
 * subscribe to String messages on a topic.
 */

#include <std_msgs/msg/string.hpp>  // Include for ROS 2 string messages
#include <string>                   // Include for std::string

#include "first_package/subscriber_interface.hpp"  // Custom subscriber node header

// Allows using the 'ms' literal for milliseconds
using namespace std::chrono_literals;

//=====================================
void SubscriberNode::receive_message(
    const std_msgs::msg::String::SharedPtr msg) {
  RCLCPP_INFO_STREAM(this->get_logger(), "Received data: " << msg->data);
}

/**
 * @brief Main function to initialize and run the ROS 2 subscriber node.
 *
 * Initializes the ROS 2 system, creates a SubscriberNode, and keeps it spinning
 * to listen for incoming messages. Shuts down the ROS 2 system before exiting.
 *
 * @param argc Number of command-line arguments.
 * @param argv Array of command-line arguments.
 * @return int Execution status code.
 */
int main(int argc, char *argv[]) {
  // Initialize the ROS2 client library for C++
  rclcpp::init(argc, argv);
  // Create an instance of SubscriberNode
  auto subscriber_node = std::make_shared<SubscriberNode>("subscriber_cpp");
  // Keep the node alive and listening for messages
  rclcpp::spin(subscriber_node);
  

  // Shut down ROS 2
  rclcpp::shutdown();
}
